src/test/regress/input/tablespace.source

```diff
@@ -146,3 +146,49 @@ DROP SCHEMA testschema CASCADE;
 
 DROP ROLE tablespace_testuser1;
 DROP ROLE tablespace_testuser2;
+
+-- Create a new tablespace and check that within this tablespace,
+-- after dropping the table, the size of all deleted (but still opened) files is 0.
+CREATE TABLESPACE testspace LOCATION '@testtablespace@';
+DROP TABLE IF EXISTS t;
+
+-- Checking that the AO row-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=row) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
+DROP TABLE t;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
+
+-- Checking that the AO column-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=column) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
+DROP TABLE t;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
+
+-- start_ignore
+-- Enable gp_inject_fault extension.
```

Comment:
how about such test optimizations?
```diff
diff --git a/src/test/regress/input/tablespace.source b/src/test/regress/input/tablespace.source
index fecc814220..61facdbadb 100755
--- a/src/test/regress/input/tablespace.source
+++ b/src/test/regress/input/tablespace.source
@@ -164,31 +164,21 @@ INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 
--- start_ignore
--- Enable gp_inject_fault extension.
--- This extension will be used in test below to force flushing of all dirty
--- pages by the bg writer process.
-CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
 -- Start with a clean state (no dirty buffers) by performing a checkpoint.
 -- It will also protect from a random timed checkpoint during the test, that could lead to not stable test result.
 CHECKPOINT;
 -- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
 -- completes its work before the actual test starts.
-SELECT pg_sleep(1);
 -- Force flushing all dirty pages by the bg writer process.
-SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
--- end_ignore
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 
 -- Checking that the temp heap table doesn't leave non-zero files.
 CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
 INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 -- Need to sleep for a second, to let the bg writer process flush the data.
-SELECT pg_sleep(1);
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 
--- start_ignore
-SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
--- end_ignore
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 
 DROP TABLESPACE testspace;
diff --git a/src/test/regress/output/tablespace.source b/src/test/regress/output/tablespace.source
index 9816a9bb78..8b75833389 100755
--- a/src/test/regress/output/tablespace.source
+++ b/src/test/regress/output/tablespace.source
@@ -276,17 +276,33 @@ INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 size: 0
+-- Start with a clean state (no dirty buffers) by performing a checkpoint.
+-- It will also protect from a random timed checkpoint during the test, that could lead to not stable test result.
+CHECKPOINT;
+-- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
+-- completes its work before the actual test starts.
+-- Force flushing all dirty pages by the bg writer process.
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+ gp_inject_fault 
+-----------------
+ Success:
+ Success:
+ Success:
+(3 rows)
+
 -- Checking that the temp heap table doesn't leave non-zero files.
 CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
 INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 -- Need to sleep for a second, to let the bg writer process flush the data.
-SELECT pg_sleep(1);
- pg_sleep 
-----------
- 
-(1 row)
-
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 size: 0
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+ gp_inject_fault 
+-----------------
+ Success:
+ Success:
+ Success:
+(3 rows)
+
 DROP TABLESPACE testspace;
```

Reply:
```
 -CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
```
removing of extension creation leads to ```ERROR:  function gp_inject_fault(unknown, unknown, smallint) does not exist``` on my side. Pretty expected I think. Does it work on your side?

```
 -- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
 -- completes its work before the actual test starts.
-SELECT pg_sleep(1);
```
Checkpointer process will call ```smgrcloseall();``` during checkpoint. If it is done during test, it will lead to test pass even if do_truncate is removed. So we need this delay to avoid this race condition.

```
  -- Need to sleep for a second, to let the bg writer process flush the data.
 -SELECT pg_sleep(1);
```

Removing sleep here can lead to always success test, even if do_truncate is removed. It is so because of race condition - without delay bg writer simply may not start data flushing before table is dropped. 
It is reflected in comments why the delay is required here.

```
 --- start_ignore
 -SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 --- end_ignore
 +SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
```

Changing gp_inject_fault_infinite to gp_inject_fault will trigger BgBufferSync only once to do flush of buffers.
If this round is complete before data insertion then test will pass even if do_truncate is removed.
gp_inject_fault_infinite will force flushing of data during all test execution. So better to keep it.

Idea of moving gp_inject_fault_infinite out of "--- start_ignore/--- end_ignore" section I do not like - because output of this command depends on cluster segment configuration.
For ex., on default 3 segments the output is:
```
--------------------------
 Success:
 Success:
 Success:
(3 rows)
```
If change to 1 segment (NUM_PRIMARY_MIRROR_PAIRS=1), the output is:
```
--------------------------
 Success:
(1 row)
```
So, to make this test case workable for any segment configuration, we need to place this function into "--- start_ignore/--- end_ignore" section.
But in tablespace test there is already another place, which relies on number of segments, so Ok, I will remove the "--- start_ignore/--- end_ignore" section.
At least it will help to catch if anything goes wrong with gp fault injection during test.

```
 --- start_ignore
 -SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 --- end_ignore
 +SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
```
Ok.

Reply:
> output of this command depends on cluster segment configuration

Almost all tests are strictly tied to the three-segment cluster configuration.